### PR TITLE
metrics-support-1

### PR DESCRIPTION
**Pull request by jgilbert01**

### CODE REVIEW

File: src/metrics/index.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        const metrics = calculateMetrics(collected);
+        // logMetrics(metrics);
+        resolve(metrics);
+      } else {
+        collected.push(x);
+        next();
+      }
+    })
+      .resume();
+  });
```
Comment (jgilbert01) [inline src/metrics/index.js:158]: - WIP
- still need to log metric via EMF
- return only count metrics from lambda for use in int tests
- might add sdk middleware to get dynamodb rcu and wcu metrics

File: test/unit/faults/index.test.js [hunk 2]
```diff
@@ -75,30 +75,29 @@ describe('faults/index.js', () => {
...
+            f1: 'v1',
+            f2: '[REDACTED]',
           },
+          eem: { fields: ['f2'] },
+        });
+        expect(collected[2].event.uow.someResponse).to.deep.equal({
+          f3: '[BUFFER: 2]',
+          f4: '[CIRCULAR]',
         });
       })
```
Comment (jgilbert01) [inline test/unit/faults/index.test.js:102]: - side effect of adding metrics to the uow
- changed all these tests to side step the new metrics element on the uow
- could impact tests of services using the lib

File: test/unit/metrics/index.test.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      .through(toPromiseWithMetrics)
+      .tap((metrics) => {
+        console.log('metrics: ');
+        console.log(JSON.stringify(metrics, null, 2));
+        // expect(collected.length).to.equal(1);
+        // expect(collected[0].updateRequest).to.deep.equal({
+        // });
+      });
+  });
+});
```
Comment (jgilbert01) [inline test/unit/metrics/index.test.js:139]: - still working on unit tests

Comment (jgilbert01) [reply test/unit/metrics/index.test.js:139]: - example metrics:
```
{
  "stream.batch.size": 6,
  "stream.batch.utilization": 0.6,
  "stream.pipeline.count": 2,
  "stream.uow.count": 5,
  "p1|stream.pipeline.utilization": 0.6,
  "p2|stream.pipeline.utilization": 0.4,
  "p1|stream.channel.wait.time": {
    "average": 2016,
    "min": 2004,
    "max": 2028,
    "sum": 6048,
    "count": 3
  },
  "p1|save|stream.pipeline.io.wait.time": {
    "average": 26,
    "min": 18,
    "max": 34,
    "sum": 78,
    "count": 3
  },
  "p1|save|stream.pipeline.io.time": {
    "average": 2,
    "min": 2,
    "max": 2,
    "sum": 6,
    "count": 3
  },
  "p1|stream.pipeline.time": {
    "average": 2060,
    "min": 2058,
    "max": 2062,
    "sum": 6180,
    "count": 3
  },
  "p2|stream.channel.wait.time": {
    "average": 2012,
    "min": 2006,
    "max": 2018,
    "sum": 4024,
    "count": 2
  },
  "p2|query|stream.pipeline.io.wait.time": {
    "average": 2,
    "min": 2,
    "max": 2,
    "sum": 4,
    "count": 2
  },
  "p2|query|stream.pipeline.io.time": {
    "average": 2,
    "min": 2,
    "max": 2,
    "sum": 4,
    "count": 2
  },
  "p2|get|stream.pipeline.io.wait.time": {
    "average": 36,
    "min": 32,
    "max": 40,
    "sum": 72,
    "count": 2
  },
  "p2|get|stream.pipeline.io.time": {
    "average": 2,
    "min": 2,
    "max": 2,
    "sum": 4,
    "count": 2
  },
  "p2|publish|stream.pipeline.io.wait.time": {
    "average": 16,
    "min": 16,
    "max": 16,
    "sum": 16,
    "count": 1
  },
  "p2|publish|stream.pipeline.io.time": {
    "average": 2,
    "min": 2,
    "max": 2,
    "sum": 2,
    "count": 1
  },
  "p2|stream.pipeline.time": {
    "average": 2073,
    "min": 2072,
    "max": 2074,
    "sum": 4146,
    "count": 2
  }
}
```

File: src/metrics/index.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  gauge(key, value) {
+    // console.log('gauge: ', key, value);
+    const k = `${this.pipeline}|${key}`;
+    this.gauges[k] = [...(this.gauges[k] || []), ...(Array.isArray(value) ? value : [value])];
+    return this;
+  }
+
+  startPipeline({ pipeline }, pipelineCount) {
```
Comment (petermyers) [inline src/metrics/index.js:52]: Maybe I'm missing something, but I don't see pipelineCount passed in to this anywhere

File: src/metrics/monitor.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      ...pipelines,
+      ...capture,
+      ...toPromise,
+    };
+  }
+
+  // could collect metrics here
+  return (event, context) => Promise.resolve().then(() => handle(event, context));
+  // could collect metrics here in .tap() and .tapCatch()
+};
```
Comment (jgilbert01) [inline src/metrics/monitor.js:20]: - add this handle wrapper in the index file
- to configure metrics
- then add env vars

File: test/unit/metrics/index.test.js [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  });
+  afterEach(() => {
+    sinon.restore();
+    mockDdb.restore();
+    mockEventBridge.restore();
+    delete process.env.BATCH_SIZE;
+    delete process.env.METRICS_ENABLED;
+  });
+
+  it('should measure pipelines', async () => {
```
Comment (codyjmoody) [inline test/unit/metrics/index.test.js:99]: When I run this test in isolation, the stream channel wait time is increased by 2 milliseconds for both p1 and p2 and the test fails.

File: src/sinks/dynamodb.js [hunk 7]
```diff
@@ -62,17 +62,17 @@ export const updateDynamoDB = ({
...
   const invoke = (uow) => {
     if (!uow[updateRequestField]) return _(Promise.resolve(uow));
 
-    const p = connector.update(uow[updateRequestField])
+    const p = connector.update(uow[updateRequestField], uow)
       .then((updateResponse) => ({ ...uow, [updateResponseField]: updateResponse }))
       .catch(rejectWithFault(uow));
 
-    return _(p); // wrap promise in a stream
+    return _(uow.metrics?.w(p, 'save') || p); // wrap promise in a stream
```
Comment (codyjmoody) [inline src/sinks/dynamodb.js:75]: would these override other checkpoints with the name 'save'? I like what you have going in the execute commands in the connectors, can we get everything from there that we may be getting here?

Comment (codyjmoody) [reply src/sinks/dynamodb.js:75]: I guess they wouldn't... the metrics test has a counter for the saves

File: src/metrics/pipelines.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  w(p, step) {
+    const self = this;
+    return new Promise((resolve, reject) => {
+      // console.log('uow: ', uow);
+      self.startStep(step);
+      return p.then(resolve, reject)
+        // TODO record capacity utilization ???
+        .tap(self.endStep(step));
+    });
+  }
```
Comment (codyjmoody) [inline src/metrics/pipelines.js:105]: I'm still working on understanding this MR, and looking at this piece in particular. If I go into cdc and duplicate the `.through(batchGetDynamoDB(rule))` line to force two batch gets, the metrics test will throw an error like `TypeError: p.then(...).tap is not a function`.  

I was mainly wondering what would happen if a single pipeline had multiple 'get' checkpoints from different steps.

Comment (jgilbert01) [reply src/metrics/pipelines.js:105]: - i would put all the gets in the same batch
- if there was a dependency i would do the 2nd gets in a 2nd pass thru the trigger
- but i'll make the step name a config option

File: test/unit/metrics/index.test.js (hunk 6 above)
Comment (jgilbert01) [reply test/unit/metrics/index.test.js:99]: - the perils of mocking Date.now()
- added Timer.now()

File: src/connectors/dynamodb.js [hunk 9]
```diff
@@ -133,44 +135,45 @@ class Connector {
...
+            return this._batchGet(unprocessed(params, response), [...attempts, response], ctx);
           } else {
             return accumlate(attempts, response);
           }
         }));
   }
 
-  _executeCommand(command) {
-    return Promise.resolve(this.db.send(command))
+  _executeCommand(command, ctx) {
```
Comment (petermyers) [inline src/connectors/dynamodb.js:174]: `_sendCommand` for consistency?

**Pull request merged**